File: packages/sdk-ts/src/resources/commerce.ts
import type { HttpClient, RequestOptions } from '../http.js';

export type CommerceEnvironment = 'sandbox' | 'live';
export type CommerceProviderKey = 'mock' | 'plural';
export type CommercePassportType = 'browse' | 'checkout';
export type CommercePaymentStatus =
  | 'created'
  | 'authorized'
  | 'checkout_created'
  | 'payment_pending'
  | 'paid'
  | 'failed'
  | 'cancelled'
  | 'expired';

export type CommerceRecord = Record<string, unknown>;

export interface CommerceDataResponse<T = CommerceRecord> {
  data: T;
  audit_event_id?: string;
  [key: string]: unknown;
}

export interface CommerceListResponse<T = CommerceRecord> {
  items: T[];
  next_cursor?: string | null;
  [key: string]: unknown;
}

export interface CommerceProfile extends CommerceRecord {
  version?: string;
  merchant?: CommerceRecord & {
    merchant_id?: string;
    display_name?: string;
    environment?: CommerceEnvironment;
    default_currency?: string;
    country_code?: string;
  };
  supported_tools?: string[];
  capabilities?: unknown[];
}

export interface CommerceIdempotentRequest {
  idempotencyKey: string;
}

export interface CommerceTenantCreateParams extends CommerceRecord {
  tenant_id?: string;
  display_name?: string;
}

export interface CommerceTenantUpdateParams extends CommerceRecord {
  display_name?: string;
  status?: 'active' | 'disabled';
}

export interface CommerceDeveloperTenantBindParams extends CommerceRecord {
  developer_id: string;
  tenant_id: string;
  role?: string;
}

export interface CommerceMerchantCreateParams extends CommerceRecord {
  display_name: string;
  category_preset: string;
  environment?: CommerceEnvironment;
  default_currency?: string;
  country_code?: string;
  agentic_commerce_requested?: boolean;
}

export interface CommerceMerchantUpdateParams extends CommerceRecord {
  display_name?: string;
  category_preset?: string;
  default_currency?: string;
  country_code?: string;
  agentic_commerce_requested?: boolean;
  agentic_commerce_enabled?: boolean;
}

export interface CommerceAgentCreateParams extends CommerceRecord {
  merchant_id: string;
  display_name: string;
  agentic_org_agent_id?: string;
  trust_status?: string;
}

export interface CommerceAgentUpdateParams extends CommerceRecord {
  display_name?: string;
  trust_status?: string;
  disabled?: boolean;
}

export interface CommerceCatalogVariantInput extends CommerceRecord {
  sku: string;
  price_amount: number;
  currency?: string;
  availability_status?: 'in_stock' | 'out_of_stock' | 'pre_order' | 'back_order' | 'unknown';
}

export interface CommerceCatalogProductCreateParams extends CommerceRecord {
  merchant_id: string;
  product_id: string;
  title: string;
  category_preset: string;
  variants: CommerceCatalogVariantInput[];
}

export interface CommerceCatalogProductUpdateParams extends CommerceRecord {
  title?: string;
  brand?: string | null;
  description?: string | null;
  image_url?: string | null;
  category_preset?: string;
  status?: 'active' | 'archived';
  variants?: CommerceRecord[];
}

export interface CommerceCatalogProductListParams {
  merchant_id?: string;
  query?: string;
  category_preset?: string;
  status?: 'active' | 'archived' | 'all';
  limit?: number;
  cursor?: string;
}

export interface CommerceCatalogSearchParams extends CommerceRecord {
  merchant_id?: string;
  query?: string;
  filters?: CommerceRecord;
  limit?: number;
  cursor?: string;
}

export interface CommerceCartCreateParams extends CommerceIdempotentRequest {
  merchant_id: string;
  currency: string;
  line_items: Array<{ variant_id: string; quantity: number }>;
  agent_id?: string;
}

export interface CommerceConsentRequestCreateParams extends CommerceRecord {
  merchant_id: string;
  passport_type: CommercePassportType;
  requested_scopes?: string[];
  max_amount?: number;
  currency?: string;
  user_principal_hint?: string;
  ttl_seconds?: number;
}

export interface CommerceConsentExchangeParams {
  consent_request_id: string;
}

export interface CommercePassportVerifyParams {
  passport_jwt: string;
  mode?: 'read_only' | 'payment_affecting';
  expected_merchant_id?: string;
}

export interface CommercePassportRevokeParams {
  jti: string;
  reason?: string;
}

export interface CommercePolicyCreateParams extends CommerceRecord {
  merchant_id: string;
  rules: CommerceRecord;
}

export interface CommercePolicyEvaluateParams extends CommerceRecord {
  merchant_id: string;
  agent_id: string;
  passport_jwt: string;
  action_scope: string;
  amount_minor_units?: number;
  currency?: string;
}

export interface CommercePaymentIntentCreateParams extends CommerceIdempotentRequest {
  merchant_id: string;
  cart_id: string;
  passport_jwt: string;
  amount_minor_units: number;
  currency: string;
  provider_key?: CommerceProviderKey;
  agent_id?: string;
  metadata?: CommerceRecord;
}

export interface CommercePaymentIntentListParams {
  merchant_id?: string;
  status?: CommercePaymentStatus;
  limit?: number;
}

export interface CommerceCheckoutLinkCreateParams extends CommerceIdempotentRequest {
  passport_jwt: string;
  success_url: string;
  cancel_url: string;
}

export interface CommerceProviderCredentialCreateParams extends CommerceRecord {
  merchant_id: string;
  provider_key: CommerceProviderKey;
  environment: CommerceEnvironment;
  credential_payload: CommerceRecord;
}

export interface CommerceProviderCredentialPatchParams extends CommerceRecord {
  credential_payload?: CommerceRecord;
  status?: 'disabled';
}

export interface CommerceProviderCredentialListParams {
  merchant_id?: string;
  provider_key?: CommerceProviderKey;
  environment?: CommerceEnvironment;
}

export interface CommerceWebhookSourceCreateParams extends CommerceRecord {
  merchant_id: string;
  source_key: string;
  display_name: string;
}

export interface CommerceWebhookSourceListParams {
  merchant_id?: string;
}

export interface CommerceOpsHealthParams {
  merchant_id?: string;
  environment?: CommerceEnvironment;
}

export interface CommerceProviderWebhookEventListParams {
  merchant_id?: string;
  processing_status?: 'received' | 'processed' | 'ignored' | 'failed';
  provider_key?: CommerceProviderKey;
  limit?: number;
}

export interface CommerceProviderWebhookReplayParams {
  reason: string;
  dry_run?: boolean;
}

export interface CommerceMcpJsonRpcRequest extends CommerceRecord {
  jsonrpc: '2.0';
  id?: string | number;
  method: string;
  params?: CommerceRecord;
}

export class CommerceClient {
  readonly #http: HttpClient;

  constructor(http: HttpClient) {
    this.#http = http;
  }

  getProfile(params?: { merchantId?: string }): Promise<CommerceProfile> {
    return this.#http.get<CommerceProfile>(
      pathWithQuery('/.well-known/grantex-commerce', { merchant_id: params?.merchantId }),
    );
  }

  mcp(request: CommerceMcpJsonRpcRequest): Promise<CommerceRecord> {
    return this.#http.post<CommerceRecord>('/mcp', request);
  }

  createTenant(params: CommerceTenantCreateParams): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>('/v1/commerce/tenants', params);
  }

  listTenants(): Promise<CommerceListResponse> {
    return this.#http.get<CommerceListResponse>('/v1/commerce/tenants');
  }

  updateTenant(tenantId: string, params: CommerceTenantUpdateParams): Promise<CommerceDataResponse> {
    return this.#http.patch<CommerceDataResponse>(
      `/v1/commerce/tenants/${encodeURIComponent(tenantId)}`,
      params,
    );
  }

  bindDeveloperTenant(params: CommerceDeveloperTenantBindParams): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>('/v1/commerce/developer-tenants', params);
  }

  createMerchant(params: CommerceMerchantCreateParams): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>('/v1/commerce/merchants', params);
  }

  getMerchant(merchantId: string): Promise<CommerceDataResponse> {
    return this.#http.get<CommerceDataResponse>(
      `/v1/commerce/merchants/${encodeURIComponent(merchantId)}`,
    );
  }

  updateMerchant(merchantId: string, params: CommerceMerchantUpdateParams): Promise<CommerceDataResponse> {
    return this.#http.patch<CommerceDataResponse>(
      `/v1/commerce/merchants/${encodeURIComponent(merchantId)}`,
      params,
    );
  }

  createAgent(params: CommerceAgentCreateParams): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>('/v1/commerce/agents', params);
  }

  listAgents(params?: { merchant_id?: string; limit?: number; cursor?: string }): Promise<CommerceListResponse> {
    return this.#http.get<CommerceListResponse>(pathWithQuery('/v1/commerce/agents', params));
  }

  getAgent(agentId: string): Promise<CommerceDataResponse> {
    return this.#http.get<CommerceDataResponse>(
      `/v1/commerce/agents/${encodeURIComponent(agentId)}`,
    );
  }

  updateAgent(agentId: string, params: CommerceAgentUpdateParams): Promise<CommerceDataResponse> {
    return this.#http.patch<CommerceDataResponse>(
      `/v1/commerce/agents/${encodeURIComponent(agentId)}`,
      params,
    );
  }

  createCatalogProduct(params: CommerceCatalogProductCreateParams): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>('/v1/commerce/catalog/products', params);
  }

  listCatalogProducts(params: CommerceCatalogProductListParams): Promise<CommerceListResponse> {
    return this.#http.get<CommerceListResponse>(pathWithQuery('/v1/commerce/catalog/products', params));
  }

  bulkUpsertCatalogProducts(params: CommerceRecord): Promise<CommerceRecord> {
    return this.#http.post<CommerceRecord>('/v1/commerce/catalog/products/bulk', params);
  }

  getCatalogProduct(productId: string, params?: { merchant_id?: string }): Promise<CommerceDataResponse> {
    return this.#http.get<CommerceDataResponse>(
      pathWithQuery(`/v1/commerce/catalog/products/${encodeURIComponent(productId)}`, params),
    );
  }

  updateCatalogProduct(
    productId: string,
    params: CommerceCatalogProductUpdateParams,
    query?: { merchant_id?: string },
  ): Promise<CommerceDataResponse> {
    return this.#http.patch<CommerceDataResponse>(
      pathWithQuery(`/v1/commerce/catalog/products/${encodeURIComponent(productId)}`, query),
      params,
    );
  }

  deleteCatalogProduct(productId: string): Promise<CommerceDataResponse> {
    return this.#http.delete<CommerceDataResponse>(
      `/v1/commerce/catalog/products/${encodeURIComponent(productId)}`,
    );
  }

  searchCatalog(params: CommerceCatalogSearchParams): Promise<CommerceListResponse> {
    return this.#http.post<CommerceListResponse>('/v1/commerce/catalog/search', params);
  }

  listAuditEvents(params?: CommerceRecord): Promise<CommerceListResponse> {
    return this.#http.get<CommerceListResponse>(pathWithQuery('/v1/commerce/audit/events', params));
  }

  createCart(params: CommerceCartCreateParams): Promise<CommerceDataResponse> {
    const { idempotencyKey, ...body } = params;
    return this.#http.post<CommerceDataResponse>(
      '/v1/commerce/carts',
      body,
      idempotencyOptions(idempotencyKey),
    );
  }

  getCart(cartId: string): Promise<CommerceDataResponse> {
    return this.#http.get<CommerceDataResponse>(`/v1/commerce/carts/${encodeURIComponent(cartId)}`);
  }

  createConsentRequest(params: CommerceConsentRequestCreateParams): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>('/v1/commerce/passports/consent-requests', params);
  }

  exchangeConsentForPassport(params: CommerceConsentExchangeParams): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>('/v1/commerce/passports/exchange', params);
  }

  listPassports(): Promise<CommerceListResponse> {
    return this.#http.get<CommerceListResponse>('/v1/commerce/passports');
  }

  verifyPassport(params: CommercePassportVerifyParams): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>('/v1/commerce/passports/verify', params);
  }

  revokePassport(params: CommercePassportRevokeParams): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>('/v1/commerce/passports/revoke', params);
  }

  createPolicy(params: CommercePolicyCreateParams): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>('/v1/commerce/policies', params);
  }

  listPolicies(params?: { merchant_id?: string; status?: string; limit?: number }): Promise<CommerceListResponse> {
    return this.#http.get<CommerceListResponse>(pathWithQuery('/v1/commerce/policies', params));
  }

  getPolicy(policyId: string): Promise<CommerceDataResponse> {
    return this.#http.get<CommerceDataResponse>(`/v1/commerce/policies/${encodeURIComponent(policyId)}`);
  }

  activatePolicy(policyId: string): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>(
      `/v1/commerce/policies/${encodeURIComponent(policyId)}/activate`,
    );
  }

  evaluatePolicy(params: CommercePolicyEvaluateParams): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>('/v1/commerce/policies/evaluate', params);
  }

  createPaymentIntent(params: CommercePaymentIntentCreateParams): Promise<CommerceDataResponse> {
    const { idempotencyKey, ...body } = params;
    return this.#http.post<CommerceDataResponse>(
      '/v1/commerce/payments/intents',
      body,
      idempotencyOptions(idempotencyKey),
    );
  }

  listPaymentIntents(params?: CommercePaymentIntentListParams): Promise<CommerceListResponse> {
    return this.#http.get<CommerceListResponse>(pathWithQuery('/v1/commerce/payments/intents', params));
  }

  getPaymentIntent(paymentIntentId: string): Promise<CommerceDataResponse> {
    return this.#http.get<CommerceDataResponse>(
      `/v1/commerce/payments/intents/${encodeURIComponent(paymentIntentId)}`,
    );
  }

  createCheckoutLink(
    paymentIntentId: string,
    params: CommerceCheckoutLinkCreateParams,
  ): Promise<CommerceDataResponse> {
    const { idempotencyKey, ...body } = params;
    return this.#http.post<CommerceDataResponse>(
      `/v1/commerce/payments/intents/${encodeURIComponent(paymentIntentId)}/checkout-link`,
      body,
      idempotencyOptions(idempotencyKey),
    );
  }

  reconcilePaymentIntent(paymentIntentId: string): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>(
      `/v1/commerce/payments/intents/${encodeURIComponent(paymentIntentId)}/reconcile`,
    );
  }

  createProviderCredential(params: CommerceProviderCredentialCreateParams): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>('/v1/commerce/provider-credentials', params);
  }

  listProviderCredentials(params?: CommerceProviderCredentialListParams): Promise<CommerceListResponse> {
    return this.#http.get<CommerceListResponse>(
      pathWithQuery('/v1/commerce/provider-credentials', params),
    );
  }

  patchProviderCredential(
    credentialId: string,
    params: CommerceProviderCredentialPatchParams,
  ): Promise<CommerceDataResponse> {
    return this.#http.patch<CommerceDataResponse>(
      `/v1/commerce/provider-credentials/${encodeURIComponent(credentialId)}`,
      params,
    );
  }

  validateProviderCredential(credentialId: string): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>(
      `/v1/commerce/provider-credentials/${encodeURIComponent(credentialId)}/validate`,
    );
  }

  createWebhookSource(params: CommerceWebhookSourceCreateParams): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>('/v1/commerce/webhook-sources', params);
  }

  listWebhookSources(params?: CommerceWebhookSourceListParams): Promise<CommerceListResponse> {
    return this.#http.get<CommerceListResponse>(pathWithQuery('/v1/commerce/webhook-sources', params));
  }

  updateWebhookSource(sourceKey: string, params: CommerceRecord): Promise<CommerceDataResponse> {
    return this.#http.patch<CommerceDataResponse>(
      `/v1/commerce/webhook-sources/${encodeURIComponent(sourceKey)}`,
      params,
    );
  }

  rotateWebhookSourceSecret(sourceKey: string): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>(
      `/v1/commerce/webhook-sources/${encodeURIComponent(sourceKey)}/rotate-secret`,
    );
  }

  getOpsHealth(params?: CommerceOpsHealthParams): Promise<CommerceRecord> {
    return this.#http.get<CommerceRecord>(pathWithQuery('/v1/commerce/ops/health', params));
  }

  listProviderWebhookEvents(params?: CommerceProviderWebhookEventListParams): Promise<CommerceListResponse> {
    return this.#http.get<CommerceListResponse>(
      pathWithQuery('/v1/commerce/ops/provider-webhook-events', params),
    );
  }

  replayProviderWebhookEvent(
    eventId: string,
    params: CommerceProviderWebhookReplayParams,
  ): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>(
      `/v1/commerce/ops/provider-webhook-events/${encodeURIComponent(eventId)}/replay`,
      params,
    );
  }

  handleProviderWebhook(
    providerKey: CommerceProviderKey,
    payload: CommerceRecord,
    options?: { headers?: Record<string, string> },
  ): Promise<CommerceDataResponse> {
    return this.#http.post<CommerceDataResponse>(
      `/v1/webhooks/providers/${encodeURIComponent(providerKey)}`,
      payload,
      options,
    );
  }
}

function idempotencyOptions(idempotencyKey: string): RequestOptions {
  return { headers: { 'Idempotency-Key': idempotencyKey } };
}

function pathWithQuery(path: string, params?: object): string {
  const query = buildQuery(params);
  return query ? `${path}?${query}` : path;
}

function buildQuery(params?: object): string {
  if (!params) return '';
  const search = new URLSearchParams();
  for (const [key, value] of Object.entries(params) as Array<[string, unknown]>) {
    if (value === undefined || value === null) continue;
    search.set(key, String(value));
  }
  return search.toString();
}

File: packages/sdk-py/src/grantex/resources/_commerce.py
from __future__ import annotations

# Commerce V1 is still preview-shaped and returns flexible JSON objects.
# The shared HTTP client returns Any, so this resource intentionally forwards
# decoded JSON dictionaries without schema coercion.
# mypy: disable-error-code=no-any-return

from typing import Any
from urllib.parse import urlencode

from .._http import HttpClient

CommerceRecord = dict[str, Any]


class CommerceClient:
    """Commerce V1/OACP resource client."""

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def get_profile(self, *, merchant_id: str | None = None) -> CommerceRecord:
        return self._http.get(_path_with_query("/.well-known/grantex-commerce", {"merchant_id": merchant_id}))

    def mcp(self, request: CommerceRecord) -> CommerceRecord:
        return self._http.post("/mcp", request)

    def create_tenant(self, params: CommerceRecord) -> CommerceRecord:
        return self._http.post("/v1/commerce/tenants", params)

    def list_tenants(self) -> CommerceRecord:
        return self._http.get("/v1/commerce/tenants")

    def update_tenant(self, tenant_id: str, params: CommerceRecord) -> CommerceRecord:
        return self._http.patch(f"/v1/commerce/tenants/{_quote(tenant_id)}", params)

    def bind_developer_tenant(self, params: CommerceRecord) -> CommerceRecord:
        return self._http.post("/v1/commerce/developer-tenants", params)

    def create_merchant(self, params: CommerceRecord) -> CommerceRecord:
        return self._http.post("/v1/commerce/merchants", params)

    def get_merchant(self, merchant_id: str) -> CommerceRecord:
        return self._http.get(f"/v1/commerce/merchants/{_quote(merchant_id)}")

    def update_merchant(self, merchant_id: str, params: CommerceRecord) -> CommerceRecord:
        return self._http.patch(f"/v1/commerce/merchants/{_quote(merchant_id)}", params)

    def create_agent(self, params: CommerceRecord) -> CommerceRecord:
        return self._http.post("/v1/commerce/agents", params)

    def list_agents(self, params: CommerceRecord | None = None) -> CommerceRecord:
        return self._http.get(_path_with_query("/v1/commerce/agents", params))

    def get_agent(self, agent_id: str) -> CommerceRecord:
        return self._http.get(f"/v1/commerce/agents/{_quote(agent_id)}")

    def update_agent(self, agent_id: str, params: CommerceRecord) -> CommerceRecord:
        return self._http.patch(f"/v1/commerce/agents/{_quote(agent_id)}", params)

    def create_catalog_product(self, params: CommerceRecord) -> CommerceRecord:
        return self._http.post("/v1/commerce/catalog/products", params)

    def list_catalog_products(self, params: CommerceRecord | None = None) -> CommerceRecord:
        return self._http.get(_path_with_query("/v1/commerce/catalog/products", params))

    def bulk_upsert_catalog_products(self, params: CommerceRecord) -> CommerceRecord:
        return self._http.post("/v1/commerce/catalog/products/bulk", params)

    def get_catalog_product(self, product_id: str, params: CommerceRecord | None = None) -> CommerceRecord:
        return self._http.get(_path_with_query(f"/v1/commerce/catalog/products/{_quote(product_id)}", params))

    def update_catalog_product(
        self,
        product_id: str,
        params: CommerceRecord,
        query: CommerceRecord | None = None,
    ) -> CommerceRecord:
        return self._http.patch(_path_with_query(f"/v1/commerce/catalog/products/{_quote(product_id)}", query), params)

    def delete_catalog_product(self, product_id: str) -> CommerceRecord:
        return self._http.delete(f"/v1/commerce/catalog/products/{_quote(product_id)}")

    def search_catalog(self, params: CommerceRecord) -> CommerceRecord:
        return self._http.post("/v1/commerce/catalog/search", params)

    def list_audit_events(self, params: CommerceRecord | None = None) -> CommerceRecord:
        return self._http.get(_path_with_query("/v1/commerce/audit/events", params))

    def create_cart(self, params: CommerceRecord, *, idempotency_key: str) -> CommerceRecord:
        return self._http.post("/v1/commerce/carts", params, headers=_idempotency_headers(idempotency_key))

    def get_cart(self, cart_id: str) -> CommerceRecord:
        return self._http.get(f"/v1/commerce/carts/{_quote(cart_id)}")

    def create_consent_request(self, params: CommerceRecord) -> CommerceRecord:
        return self._http.post("/v1/commerce/passports/consent-requests", params)

    def exchange_consent_for_passport(self, params: CommerceRecord) -> CommerceRecord:
        return self._http.post("/v1/commerce/passports/exchange", params)

    def list_passports(self) -> CommerceRecord:
        return self._http.get("/v1/commerce/passports")

    def verify_passport(self, params: CommerceRecord) -> CommerceRecord:
        return self._http.post("/v1/commerce/passports/verify", params)

    def revoke_passport(self, params: CommerceRecord) -> CommerceRecord:
        return self._http.post("/v1/commerce/passports/revoke", params)

    def create_policy(self, params: CommerceRecord) -> CommerceRecord:
        return self._http.post("/v1/commerce/policies", params)

    def list_policies(self, params: CommerceRecord | None = None) -> CommerceRecord:
        return self._http.get(_path_with_query("/v1/commerce/policies", params))

    def get_policy(self, policy_id: str) -> CommerceRecord:
        return self._http.get(f"/v1/commerce/policies/{_quote(policy_id)}")

    def activate_policy(self, policy_id: str) -> CommerceRecord:
        return self._http.post(f"/v1/commerce/policies/{_quote(policy_id)}/activate")

    def evaluate_policy(self, params: CommerceRecord) -> CommerceRecord:
        return self._http.post("/v1/commerce/policies/evaluate", params)

    def create_payment_intent(self, params: CommerceRecord, *, idempotency_key: str) -> CommerceRecord:
        return self._http.post("/v1/commerce/payments/intents", params, headers=_idempotency_headers(idempotency_key))

    def list_payment_intents(self, params: CommerceRecord | None = None) -> CommerceRecord:
        return self._http.get(_path_with_query("/v1/commerce/payments/intents", params))

    def get_payment_intent(self, payment_intent_id: str) -> CommerceRecord:
        return self._http.get(f"/v1/commerce/payments/intents/{_quote(payment_intent_id)}")

    def create_checkout_link(
        self,
        payment_intent_id: str,
        params: CommerceRecord,
        *,
        idempotency_key: str,
    ) -> CommerceRecord:
        return self._http.post(
            f"/v1/commerce/payments/intents/{_quote(payment_intent_id)}/checkout-link",
            params,
            headers=_idempotency_headers(idempotency_key),
        )

    def reconcile_payment_intent(self, payment_intent_id: str) -> CommerceRecord:
        return self._http.post(f"/v1/commerce/payments/intents/{_quote(payment_intent_id)}/reconcile")

    def create_provider_credential(self, params: CommerceRecord) -> CommerceRecord:
        return self._http.post("/v1/commerce/provider-credentials", params)

    def list_provider_credentials(self, params: CommerceRecord | None = None) -> CommerceRecord:
        return self._http.get(_path_with_query("/v1/commerce/provider-credentials", params))

    def patch_provider_credential(self, credential_id: str, params: CommerceRecord) -> CommerceRecord:
        return self._http.patch(f"/v1/commerce/provider-credentials/{_quote(credential_id)}", params)

    def validate_provider_credential(self, credential_id: str) -> CommerceRecord:
        return self._http.post(f"/v1/commerce/provider-credentials/{_quote(credential_id)}/validate")

    def create_webhook_source(self, params: CommerceRecord) -> CommerceRecord:
        return self._http.post("/v1/commerce/webhook-sources", params)

    def list_webhook_sources(self, params: CommerceRecord | None = None) -> CommerceRecord:
        return self._http.get(_path_with_query("/v1/commerce/webhook-sources", params))

    def update_webhook_source(self, source_key: str, params: CommerceRecord) -> CommerceRecord:
        return self._http.patch(f"/v1/commerce/webhook-sources/{_quote(source_key)}", params)

    def rotate_webhook_source_secret(self, source_key: str) -> CommerceRecord:
        return self._http.post(f"/v1/commerce/webhook-sources/{_quote(source_key)}/rotate-secret")

    def get_ops_health(self, params: CommerceRecord | None = None) -> CommerceRecord:
        return self._http.get(_path_with_query("/v1/commerce/ops/health", params))

    def list_provider_webhook_events(self, params: CommerceRecord | None = None) -> CommerceRecord:
        return self._http.get(_path_with_query("/v1/commerce/ops/provider-webhook-events", params))

    def replay_provider_webhook_event(self, event_id: str, params: CommerceRecord) -> CommerceRecord:
        return self._http.post(f"/v1/commerce/ops/provider-webhook-events/{_quote(event_id)}/replay", params)

    def handle_provider_webhook(
        self,
        provider_key: str,
        payload: CommerceRecord,
        *,
        headers: dict[str, str] | None = None,
    ) -> CommerceRecord:
        return self._http.post(f"/v1/webhooks/providers/{_quote(provider_key)}", payload, headers=headers)


def _idempotency_headers(idempotency_key: str) -> dict[str, str]:
    return {"Idempotency-Key": idempotency_key}


def _path_with_query(path: str, params: CommerceRecord | None = None) -> str:
    if not params:
        return path
    query = urlencode({k: v for k, v in params.items() if v is not None})
    return f"{path}?{query}" if query else path


def _quote(value: str) -> str:
    from urllib.parse import quote

    return quote(value, safe="")
